core/src/main/java/io/grpc/internal/StatsTraceContext.java

```diff
@@ -0,0 +1,229 @@
+    this.censusContext = parentContext.with(methodTagKey, new TagValue(fullMethodName));
+    this.stopwatch = stopwatchSupplier.get().start();
+    this.censusHeader = censusHeader;
+  }
+
+  /**
+   * Creates a {@code StatsTraceContext} for an outgoing RPC, using the current CensusContext.
+   *
+   * <p>The current time is used as the start time of the RPC.
+   */
+  public static StatsTraceContext newClientContext(String methodName,
+      CensusContextFactory censusFactory, Supplier<Stopwatch> stopwatchSupplier) {
+    return new StatsTraceContext(Side.CLIENT, methodName,
+        // TODO(zhangkun83): use the CensusContext out of the current Context
+        censusFactory.getDefault(),
+        stopwatchSupplier, createCensusHeader(censusFactory));
+  }
+
+  @VisibleForTesting
+  static StatsTraceContext newClientContextForTesting(String methodName,
+      CensusContextFactory censusFactory, CensusContext parent,
+      Supplier<Stopwatch> stopwatchSupplier) {
+    return new StatsTraceContext(Side.CLIENT, methodName, parent, stopwatchSupplier,
+        createCensusHeader(censusFactory));
+  }
+
+  /**
+   * Creates a {@code StatsTraceContext} for an incoming RPC, using the CensusContext deserialized
+   * from the headers.
+   *
+   * <p>The current time is used as the start time of the RPC.
+   */
+  public static StatsTraceContext newServerContext(String methodName,
+      CensusContextFactory censusFactory, Metadata headers,
+      Supplier<Stopwatch> stopwatchSupplier) {
+    Metadata.Key<CensusContext> censusHeader = createCensusHeader(censusFactory);
+    CensusContext parentContext = headers.get(censusHeader);
+    if (parentContext == null) {
+      parentContext = censusFactory.getDefault();
+    }
+    return new StatsTraceContext(Side.SERVER, methodName, parentContext, stopwatchSupplier,
+        censusHeader);
+  }
+
+  /**
+   * Propagate the context to the outgoing headers.
+   */
+  void propagateToHeaders(Metadata headers) {
+    headers.discardAll(censusHeader);
+    headers.put(censusHeader, censusContext);
+  }
+
+  Metadata.Key<CensusContext> getCensusHeader() {
+    return censusHeader;
+  }
+
+  @VisibleForTesting
+  CensusContext getCensusContext() {
+    return censusContext;
+  }
+
+  @VisibleForTesting
+  static Metadata.Key<CensusContext> createCensusHeader(
+      final CensusContextFactory censusContextFactory) {
+    return Metadata.Key.of("grpc-census-bin", new Metadata.BinaryMarshaller<CensusContext>() {
+        @Override
+        public byte[] toBytes(CensusContext context) {
+          return context.serialize().array();
+        }
+
+        @Override
+        public CensusContext parseBytes(byte[] serialized) {
+          return censusContextFactory.deserialize(ByteBuffer.wrap(serialized));
+        }
+      });
+  }
+
+  /**
+   * Record the outgoing number of payload bytes as on the wire.
+   */
+  void wireBytesSent(long bytes) {
+    checkState(!callEnded, "already eneded");
```

Comment:
This is a lot of state checking for something that probably isn't that bad.  You should probably consider reducing them to asserts after this is stable.